Optimize decommit

PeterSolMS opened the pull request:
Optimize decommitting GC heap memory pages:

1. Move the decommit out of the GC pause for server GC
2. Throttle decommit so the decommit rate stays reasonable
3. Take gen 1 budget and free list space into account when planning what to decommit
4. Add exponential smoothing for the decommit target to avoid decommit/recommit for short term fluctuations

PeterSolMS committed:
Initial snap shot of decommit optimization code.

Key ideas:
 - in decommit_ephemeral_segment_pages, instead of decommitting right there, just remember what we intended to decommit.
- in gc_thread function, have the thread with heap_number 0 periodically wake up and do gradual decommits for all heaps by calling the new method decommit_ephemeral_pages_step on them.
- decommit_ephemeral_pages_step will decommit a "reasonable" amount of space per heap and return how much it decomitted. A return value of 0 indicates no further progress.

PeterSolMS committed:
Fixed issues with the gradual decommit logic.

Details:
- add code checking the assumption that nobody is changing the committed size or the desired allocation size while we do the decommitting
- fix logic error causing race between committing and decommitting
- add a decommit step at the end of GC to make sure decommit would still happen even for very frequent GCs.
- fix issue that crept in during pinned heap work

PeterSolMS committed:
Refinement to decommit policy taking gen1 budget and free list space into account.

PeterSolMS committed:
Move consideration of loh_threshold to decommit_ephemeral_segment_pages.

Fix bug in accounting for free list space.

PeterSolMS committed:
Merge branch 'master' into Optimize-decommit

PeterSolMS committed:
Streamline logic in decommit_ephemeral_segment_pages, add exponential smoothing of the decommit target when ramping down, limit decommit rate for workstation GC.

PeterSolMS committed:
Merge branch 'master' into Optimize-decommit

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+        const size_t DECOMMIT_STEP_SIZE = max (4096 / n_heaps, 100) * OS_PAGE_SIZE;
+        uint8_t* page_start = align_on_page ( max (decommit_target, committed - DECOMMIT_STEP_SIZE) );
+        size_t size = committed - page_start;
+        virtual_decommit(page_start, size, heap_number);
+        dprintf(3, ("Decommitting heap segment [%Ix, %Ix[(%d)",
+            (size_t)page_start,
+            (size_t)(page_start + size),
+            size));
+        heap_segment_committed (ephemeral_heap_segment) = page_start;
+        ephemeral_heap_segment->saved_committed = page_start;
+        if (heap_segment_used (ephemeral_heap_segment) > heap_segment_committed (ephemeral_heap_segment))
+        {
+            heap_segment_used (ephemeral_heap_segment) = heap_segment_committed (ephemeral_heap_segment);
```
this is duplicating part of the code in `decommit_heap_segment_pages`; would be nice to have a worker function to do the common part so you don't need to care about details like updating used here. it's better to have that taken care of than having the person who's looking at this method later having to care about used.

PeterSolMS replied:
Done - common part is called decommit_heap_segment_pages_worker.

Maoni0 commented on src/coreclr/src/gc/gcpriv.h:
```diff
@@ -4695,7 +4702,10 @@ class heap_segment
     uint8_t*        background_allocated;
 #ifdef MULTIPLE_HEAPS
     gc_heap*        heap;
+    uint8_t*        saved_committed;
+    size_t          saved_desired_allocation;
```
I would make these _DEBUG only fields and assert on them - it's nice to not have to increase the size of `heap_segment` in release and these things really should be guarantees so asserts are more appropriate.

PeterSolMS replied:
Done.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+            GCToOSInterface::DebugBreak();
+        }
+
+        const size_t DECOMMIT_STEP_SIZE = max (4096 / n_heaps, 100) * OS_PAGE_SIZE;
```
instead of doing this calculation here, I would suggest to make 160k a define and use a multiple of that so it's clear how much time you are expressing here for each decommit step. perhaps 4x 160k?

PeterSolMS replied:
I used defines for both the time step in milliseconds and the decommit size per millisecond.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+#ifndef MULTIPLE_HEAPS
+    // we want to limit the amount of decommit we do per time to indirectly
+    // limit the amount of time spent in recommit and page faults
+    size_t ephemeral_elapsed = dd_time_clock (dd0) - gc_last_ephemeral_decommit_time;
```
you can get rid of `GC_EPHEMERAL_DECOMMIT_TIMEOUT` now you are using it anymore.

PeterSolMS replied:
Yes - done.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+    ptrdiff_t decommit_size = heap_segment_committed (ephemeral_heap_segment) - decommit_target;
+
+    // we do a max of 160 kB per millisecond (160 MB per second) of pause time
+    // we don't want to take large pause times too seriously - limit to 10 seconds
```
I think this comment can sound very misleading - we definitely want to take large pause time very seriously 😄 I think you meant to say the amount of decommit we want to do is kind of proportional to the elapsed time inbetween GCs, as in, if there's only a little time elapsed we wouldn't want to spend a lot of paused doing decommit. perhaps say that instead.

PeterSolMS replied:
Fixed the comment to make it clearer and avoid the possible misunderstanding.

Maoni0 reviewed (changes requested):
other than those comments I just have a nit comment on the coding style - it appears the spacing is done randomly. in GC we (well I at least 😄) try to conform to the style that is 

one space between method/macro and '(', and no space if the method/macro takes no argument. 

however I've been thinking to eliminate the 2nd part since it just seems folks in general have trouble conforming to it. instead of having to tell everyone a longer rule I'd rather tell them a shorter rule. so I'm ok if you leave the space in `gc_heap::decommit_step ()`. at some point we can convert all of them to having one space. 

also there should always be () around an expression, even when it's simple just so that we are constant so this line 

`uint8_t* page_start = align_on_page ( max (decommit_target, committed - DECOMMIT_STEP_SIZE) );`

should be changed to 

`uint8_t* page_start = align_on_page (max (decommit_target, (committed - DECOMMIT_STEP_SIZE)));`

and 

`heap_segment_committed(ephemeral_heap_segment)`

should be changed to 

`heap_segment_committed (ephemeral_heap_segment)`

thanks for conforming to these to make our code base more consistent :)

Maoni0 commented:
for testing, please make sure the perf testing is done as we discussed, thanks! I would also do a stress run on this (I'm paranoid) even though I don't think it would turn up anything.

PeterSolMS committed:
Fix 32-bit compile error

PeterSolMS committed:
Partially address code review feedback.

PeterSolMS committed:
Address code review feedback, avoid possible overflow in address computation.

PeterSolMS committed:
Add spaces in a couple places before opening parentheses.

PeterSolMS commented:
I also fixed the coding style issues.

Maoni0 commented on src/coreclr/src/gc/gc.cpp:
```diff
@@ -31727,56 +31752,137 @@ void gc_heap::trim_youngest_desired_low_memory()
+        decommit_size = max (decommit_size, (ptrdiff_t)(100 * OS_PAGE_SIZE));
+
+        // but don't try to do more than what's actually committed in the segment
+        decommit_size = min ((heap_segment_committed (ephemeral_heap_segment) - heap_segment_mem (ephemeral_heap_segment)),
```
this should be
```
decommit_size = min ((heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment)),
```
however I think we should be able to refactor this to make it simpler... the changes involve so many different numbers when calculating the decommit size, plus the existing ones... that's a lot to keep track of. I'll have some suggestion to simplify it some.

PeterSolMS replied:
Your code works too, my intention was just to guard against the case of a large decommit size and a very low segment address where committed - decommit_size would overflow and wrap around.

Maoni0 commented:
what do you think of this?

```c++
// get rid of decommit_heap_segment_pages_worker
size_t gc_heap::decommit_heap_segment_pages (heap_segment* seg,
                                             uint8_t* new_committed=0)
{
    if (use_large_pages_p)
        return;
    // Note that I got the extra 4 pages that you had in decommit_ephemeral_segment_pages_step; 
    // If there is a reason that it's 4 and not 2, you could change the 2 pages here to 4.
    uint8_t* page_start = align_on_page (new_committed ? new_committed : heap_segment_allocated (seg)) + (2 * OS_PAGE_SIZE);
    size_t size = 0;
    if (page_start < heap_segment_committed (seg))
    {
        size_t size = heap_segment_committed (seg) - page_start;
        if (size > (100 * OS_PAGE_SIZE))
        {
            virtual_decommit (page_start, size, heap_number);
            dprintf (3, ("Decommitting heap segment [%Ix, %Ix[(%d)",
                (size_t)page_start,
                (size_t)(page_start + size),
                size));
            heap_segment_committed (seg) = page_start;
            if (heap_segment_used (seg) > heap_segment_committed (seg))
            {
                heap_segment_used (seg) = heap_segment_committed (seg);
            }
        }
        else
            // didn't do any decommit, return 0.
            size = 0;
    }

    return size;
}


size_t gc_heap::decommit_ephemeral_segment_pages_step ()
{
    // we rely on desired allocation not being changed outside of GC
    assert (ephemeral_heap_segment->saved_desired_allocation == dd_desired_allocation (dynamic_data_of (0)));

    uint8_t* decommit_target = heap_segment_decommit_target (ephemeral_heap_segment);
    uint8_t* committed = heap_segment_committed (ephemeral_heap_segment);
    if (decommit_target < committed)
    {
        // we rely on other threads not messing with committed if we are about to trim it down
        assert (ephemeral_heap_segment->saved_committed == heap_segment_committed (ephemeral_heap_segment));

        // limit the decommit size to some reasonable value per time interval
        ptrdiff_t decommit_size = ((DECOMMIT_SIZE_PER_MILLISECOND * DECOMMIT_TIME_STEP_MILLISECONDS) / n_heaps);

        // figure out where the new committed should be
        uint8_t* new_committed = max (decommit_target, (committed - decommit_size));
        size_t size = decommit_heap_segment_pages (ephemeral_heap_segment, new_committed);

#ifdef _DEBUG
        ephemeral_heap_segment->saved_committed = committed - size;
#endif // _DEBUG

        return size;
    }
    return 0;
}
```
the `decommit_heap_segment_pages` we are calling for WKS will need to pass in an appropriate new_committed which is `(allocated + slack_space)`.

PeterSolMS commented:
The problem with your suggested solution is that the additional 2 pages and the 100 page threshold in decommit_heap_segment_pages will apply to every decommit step. So if you have more than about 40 heaps, nothing will actually happen in decommit_heap_segment_pages when called by decommit_ephemeral_segment_pages_step because the step is smaller than 100 pages. 

Of course you could just compensate for this in decommit_ephemeral_segment_pages_step by asking for at least 103 pages, but this is error prone because decommit_ephemeral_segment_pages_step would now knows what's going on in decommit_heap_segment_pages, and thus if one is changed but not the other, the code will malfuntion.

I also think the complications I put in make sense (or at least we should discuss in detail):
- on machines with many heaps we should have a minimum step per heap (whether it should be a 100 pages or something else is of course debatable).
- I think it also makes sense to guard against arithmetic overflow in the subtraction "committed - decommit_size".

PeterSolMS committed:
Addressed code review feedback about the computation of decommit_size in decommit_ephemeral_pages_step.

Maoni0 commented:
ahh, I misread then. I thought you also wanted 100pages minimum but now I read your last iteration and clearly that was not the case. this was the policy we had in the implementation before you change. I actually don't think 100 pages are that bad. but I'm also fine with fewer pages when the # of heaps is large. in which we can just change the 100 number and still simplify it because this new logic + the old logic is unnecessarily complicated.

would you be ok if we changed 100 to something like 20 pages (~ a large object size)? 

> I think it also makes sense to guard against arithmetic overflow in the subtraction "committed - decommit_size".

I noticed that too and the implementation I suggested does not have this problem.

PeterSolMS commented:
> I noticed that too and the implementation I suggested does not have this problem.
Could you explain why not? I think it *does* have this problem in the case of a segment with a very low virtual address and very little committed memory.

The thing that bothers me about your solution is the implied coupling between decommit_ephemeral_segment_pages_step and decommit_heap_segment_pages in the sense that the former has to know about the internal workings of the latter. That's why I thought it cleaner to have a worker method has the common core that just does the mechanics exactly as it is being told.

PeterSolMS committed:
Fixes to decommit changes:
- move computation of reasonable decommit step size to init code
- change value of EXTRA_SPACE in decommit_ephemeral_segment_pages_step to be consistent with decommit_heap_segment_pages
- reformulate computation of new_committed in decommit_ephemeral_segment_pages_step to eliminate possibility of arithmetic overflow.

Maoni0 reviewed (approved):
LGTM!

PeterSolMS committed:
Create a #define MIN_DECOMMIT_SIZE for the 100*OS_PAGE minimum decommit we do per OS call.
